File: src/main/java/factory/ViewFactory.java
package factory;

import javafx.fxml.FXMLLoader;
import javafx.fxml.Initializable;
import javafx.scene.Node;

import java.io.IOException;

//invariant: controller is set only after calling load() method.
public class ViewFactory<C extends Initializable> {

    private String resourcePath;
    protected C controller = null;

    public ViewFactory(String resourcePath){
        this.resourcePath = resourcePath;
    }

    public Node load() throws IOException{
        FXMLLoader fxmlLoader = new FXMLLoader(getClass().getClassLoader().getResource(resourcePath));

        Node n = fxmlLoader.load();
        controller = fxmlLoader.getController();

        return n;
    }

    public Node load(C controller) throws IOException{
        FXMLLoader fxmlLoader = new FXMLLoader(getClass().getClassLoader().getResource(resourcePath));
        fxmlLoader.setController(controller);

        Node n = fxmlLoader.load();
        this.controller = controller;

        return n;
    }

    public C getController() { return controller; }
}

File: src/main/scala/factory/ViewFactory.scala
package factory

import java.io.{File, IOException}

import javafx.fxml.{FXMLLoader, Initializable}
import javafx.scene.Node

//invariant: controller will only be set after calling load method
class ViewFactory[C <: Initializable](resourcePath: String) {

    protected var controller: C = _

    @throws[IOException]
    def load: Node = {
        val fxmlLoader = new FXMLLoader(new File(resourcePath).toURI.toURL)

        val n = fxmlLoader.load
        controller = fxmlLoader.getController[C]

        n
    }

    @throws[IOException]
    def load(controller: C): Node = {
        val fxmlLoader = new FXMLLoader(new File(resourcePath).toURI.toURL)
        fxmlLoader.setController(controller)

        val n = fxmlLoader.load[Node]
        this.controller = controller

        n
    }

    def getController: C = controller
}

